### ISSUE TITLE
Questions about the performance of the Neural ODE example

### ISSUE THREAD

**Issue by zhlfmyzhh** [patrick-kidger/diffrax#671]
I am a new beginner of jax and diffrax. I find the notebook example is clear, but the performance seems not to be not good. The neural ODE can not entirely fit the curve. I've tried different activation functions such as swish, more depths and width size for the MLP, smaller stepsize_controller error, different steps_strategy, length_strategy and cosine decay learning rate sceduler. However, the model can still not entirely fit the curve. 

I wonder it's neural ODE's limit or anything else I can do to further improve it? Thank you very much for your time and help!

**Comment by zhlfmyzhh** [comment]
I tried different ways. For example, I try 
1. Activation function:   softplus
2. Step strategy:   (500, 500, 500, 500, 500, 500, 500, 500, 500, 500)
3. Length strategy:   (0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0)
4. Learning rate scheduler
   "init": 1e-4,
   "peak": 3e-3,
   "warmup": 1200,
   "decay": 3000
   "end": 1e-5
5. solver:   rtol=1e-5, atol=1e-7
6. width_size=64, depth=8
7. Smaller time:   ts = jnp.linspace(0, 5, 200)

I have also tried different examples, such as 
        dy1 = y2
        dy2 = -jnp.sin(y1) - 0.1 * y2
Similar phenomenons are observed.

**Comment by patrick-kidger** [comment]
Thanks for the question! So I'll note that the Neural ODE example already fits the provided data very closely indeed. Getting below this is likely to require a very small learning rate, for example.

For what it's worth you have just prompted me to write #672, which tweaks the example to use a missing best-practice around `tanh` as a final activation function.

**Comment by zhlfmyzhh** [comment]
Thank you for your prompt reply and help! I found that for your new example, your training output and the figure do not match the code. As your seed is fixed, it should be 

```
Step: 0, Loss: 0.1405617743730545, Computation time: 5.052995204925537
Step: 100, Loss: 0.01317338552325964, Computation time: 0.023674488067626953
Step: 200, Loss: 0.009179404936730862, Computation time: 0.023085355758666992
Step: 300, Loss: 0.0021185341756790876, Computation time: 0.02717304229736328
Step: 400, Loss: 0.0015776436775922775, Computation time: 0.032082319259643555
Step: 499, Loss: 0.002610038500279188, Computation time: 0.020656585693359375
Step: 0, Loss: 0.06953904032707214, Computation time: 4.626847267150879
Step: 100, Loss: 0.02050553821027279, Computation time: 0.10858273506164551
Step: 200, Loss: 0.017762290313839912, Computation time: 0.11797380447387695
Step: 300, Loss: 0.0074322340078651905, Computation time: 0.1298081874847412
Step: 400, Loss: 0.005885579623281956, Computation time: 0.13971590995788574
Step: 499, Loss: 0.003939368762075901, Computation time: 0.13250136375427246
Final average MSE across dataset: 0.004574755
```

The loss and the figure should be the same. Final average MSE is some snippets written by myself so as to examine the performance.

<img width="1084" height="806" alt="Image" src="https://github.com/user-attachments/assets/944d82c3-945e-43e3-9a00-c1a02080a028" />

```
def compute_mse(model, ts, ys):
    y0s = ys[:, 0]
    preds = jax.vmap(model, in_axes=(None, 0))(ts, y0s)  # shape: (N, T, D)
    return jnp.mean((preds - ys) ** 2)

final_mse = compute_mse(model, ts, ys)
print("Final average MSE across dataset:", final_mse)
```

Besides, I found that your original code has a better performance than the new one.
```
Step: 0, Loss: 0.05405808612704277, Computation time: 4.508039474487305
Step: 100, Loss: 0.0009977879235520959, Computation time: 0.017888545989990234
Step: 200, Loss: 0.0006626047543250024, Computation time: 0.023993968963623047
Step: 300, Loss: 0.00014425546396523714, Computation time: 0.02193617820739746
Step: 400, Loss: 6.219919305294752e-05, Computation time: 0.020198345184326172
Step: 499, Loss: 6.944094639038667e-05, Computation time: 0.02062845230102539
Step: 0, Loss: 0.0014205964980646968, Computation time: 4.514024496078491
Step: 100, Loss: 0.0002869167074095458, Computation time: 0.11824464797973633
Step: 200, Loss: 0.0003324180142953992, Computation time: 0.12994933128356934
Step: 300, Loss: 0.0002099012490361929, Computation time: 0.1452326774597168
Step: 400, Loss: 0.0001296987320529297, Computation time: 0.14360857009887695
Step: 499, Loss: 0.00020818821212742478, Computation time: 0.1065073013305664
Final average MSE across dataset: 0.00027261826
```

After that, I try to change `softplus` to `swish` and reduce `rtol=1e-4, atol=1e-7`. The performance is better, but it is still a bit worse than the original version. May you give me some suggestions?

**Comment by patrick-kidger** [comment]
> Thank you for your prompt reply and help! I found that for your new example, your training output and the figure do not match the code. As your seed is fixed, it should be

If you mean that you're getting slightly different results then this is surprising, as I simply ran the notebook and saved it. If there is a difference it is presumably due to machine-to-machine differences of some flavour, a different version of JAX, etc.

> Besides, I found that your original code has a better performance than the new one.

This I'm not really worried by - the code is intended to be illustrative of best practices, not carefully tuned to get a minimal loss. (This would just be overfitting to a toy problem anyway.)

> After that, I try to change softplus to swish and reduce rtol=1e-4, atol=1e-7. The performance is better, but it is still a bit worse than the original version. May you give me some suggestions?

The toy problem in question is already solved very closely by the provided example. I'm sure you could try adjusting hyperparameters to get it even closer (using a small learning rate with pure SGD would be one common way of doing this), if that was really desired. But for all practical problems this would already be considered 'solved', so I don't have too many suggestions :)

**Comment by zhlfmyzhh** [comment]
Thank you for your prompt clarification and your amazing work and efforts! Thank you again!
